hail/python/hailtop/batch_client/aioclient.py

```diff
@@ -410,21 +414,20 @@ async def delete(self):
         await self._client._delete(f'/api/v1alpha/batches/{self.id}')
 
 
-class BatchBuilder:
-    def __init__(self, client, attributes, callback, token=None, cancel_after_n_failures=None):
-        self._client = client
-        self._job_idx = 0
-        self._job_specs = []
-        self._jobs = []
-        self._submitted = False
-        self.attributes = attributes
-        self.callback = callback
+class BatchBuilder(abc.ABC):
+    _submitted: bool
+    _job_idx: int
+    _jobs: List[Job]
+    _job_specs: List[dict]
+    _client: 'BatchClient'
```

Comment:
I think this is better done with a constructor and the deriving class calling `super().__init__()`. I believe these fields are [shared by all instances of the class](https://stackoverflow.com/questions/1537202/difference-between-variables-inside-and-outside-of-init).